presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -89,400 +111,733 @@
+                node.getWhere(),
+                node.getGroupBy(),
+                node.getHaving(),
+                node.getOrderBy(),
                 node.getOffset(),
                 node.getLimit());
     }
 
     @Override
-    protected Node visitSelect(Select node, Void context)
+    protected Node visitNode(Node node, Void context)
     {
-        if (materializedViewInfo.isDistinct() && !node.isDistinct()) {
-            throw new IllegalStateException("Materialized view has distinct and base query does not");
-        }
-        ImmutableList.Builder<SelectItem> rewrittenSelectItems = ImmutableList.builder();
-
-        for (SelectItem selectItem : node.getSelectItems()) {
-            rewrittenSelectItems.add((SelectItem) process(selectItem, context));
-        }
-
-        return new Select(node.isDistinct(), rewrittenSelectItems.build());
+        return node;
     }
 
     @Override
-    protected Node visitSingleColumn(SingleColumn node, Void context)
+    protected Node visitQuery(Query node, Void context)
     {
-        // For a single table, without sub-queries, the column prefix is unnecessary. Here It is removed so that it can be mapped to the view column properly.
-        // For relations other than single table, it needs to be reserved to differentiate columns from different tables.
-        // One way to do so is to process the prefix within `visitDereferenceExpression()` since the prefix information is saved as `base` in `DereferenceExpression` node.
-        node = removeSingleColumnPrefix(node, removablePrefix);
-        Expression expression = node.getExpression();
-        Optional<Set<Expression>> groupByOfMaterializedView = materializedViewInfo.getGroupBy();
-        // TODO: Replace this logic with rule-based validation framework.
-        if (groupByOfMaterializedView.isPresent() &&
-                validateExpressionForGroupBy(groupByOfMaterializedView.get(), expression) &&
-                (!expressionsInGroupBy.isPresent() || !expressionsInGroupBy.get().contains(expression))) {
-            throw new IllegalStateException("Query a column presents in materialized view group by: " + expression.toString());
-        }
-
-        Expression processedColumn = (Expression) process(expression, context);
-        Optional<Identifier> alias = node.getAlias();
-
-        // If a column name was rewritten, make sure we re-alias to same name as base query
-        if (!alias.isPresent() && processedColumn instanceof Identifier && !processedColumn.equals(node.getExpression())) {
-            alias = Optional.of((Identifier) node.getExpression());
-        }
-        return new SingleColumn(processedColumn, alias);
+        QueryBody newQueryBody = processSameType(node.getQueryBody());
+        Optional<With> newWith = node.getWith().map(this::processSameType);
+        boolean withSame = !node.getWith().isPresent() || (node.getWith().orElseThrow(IllegalStateException::new) == newWith.get());
```

Comment:
Add detail message for the illegal state.

Reply:
TBH I don't believe we should even be using `orElseThrow` here- we are immediately preceding it with a check to `isPresent()`. I replaced with `get()` but am open to further discussion.